src/lazy/expanded/compiler.rs

```diff
@@ -272,18 +272,128 @@ impl TemplateCompiler {
-        match value.annotations().next() {
-            None => Ok(ParameterEncoding::Tagged),
-            Some(Ok(text)) => match text.expect_text()? {
+    /// Interprets the annotations on the parameter name to determine its encoding.
+    fn encoding_for<Encoding: Decoder>(
+        context: EncodingContextRef,
+        pending_macros: &MacroTable,
+        parameter: LazyValue<Encoding>,
+    ) -> IonResult<ParameterEncoding> {
+        // * If the parameter has no annotations, it uses the default encoding.
+        //   For example:
+        //       foo
+        // * If the parameter has one annotation, the annotation is the encoding name. This can be
+        //   either a built-in encoding like `uint8` or a user-defined macro that takes at least
+        //   one parameter of its own.
+        //   For example:
+        //          uint8::foo
+        //        float32::foo
+        //       my_macro::foo
+        // * If the parameter has two annotations, the first annotation is the module name and the
+        //   second is the name of an encoding in that module.
+        //   For example:
+        //               $ion::uint8::foo
+        //             $ion::float32::foo
+        //       my_module::my_macro::foo
+        let mut annotations_iter = parameter.annotations();
+        let annotation1 = match annotations_iter.next().transpose()? {
+            // If there aren't any annotations, this parameter uses the default encoding: tagged.
+            None => return Ok(ParameterEncoding::Tagged),
+            // Otherwise, it's a module name or encoding name. We'll have to see if there's a
+            // second annotation to know for certain. Either way, the annotation must have known text.
+            Some(symbol) => symbol.expect_text()?,
+        };
+        let annotation2 = match annotations_iter.next().transpose()? {
+            // If there's a second annotation, get its associated text.
+            Some(symbol) => Some(symbol.expect_text()?),
+            None => None,
+        };
+
+        // Make sure there isn't a third annotation.
+        match annotations_iter.next() {
+            // This is the expected case. Having more than 2 annotations is illegal.
+            None => {}
+            Some(Err(e)) => return Err(e),
+            Some(Ok(annotation)) => {
+                return IonResult::decoding_error(format!(
+                    "found unexpected third annotation ('{:?}') on parameter",
+                    annotation
+                ))
+            }
+        };
+
+        // If it's an unqualified name, look for the associated macro in the local scope.
+        if let (encoding_name, None) = (annotation1, annotation2) {
+            if let Some(macro_ref) =
+                Self::resolve_unqualified_macro_id(context, pending_macros, encoding_name)
+            {
+                Self::validate_macro_shape_for_encoding(&macro_ref)?;
```

Comment:
🗺️ This check makes sure the macro isn't a constant. Constants cannot be macro-shaped parameter encodings.